src/kernel/arch/x86/task.zig

```diff
@@ -0,0 +1,265 @@
+const std = @import("std");
+const expectEqual = std.testing.expectEqual;
+const expectError = std.testing.expectError;
+const assert = std.debug.assert;
+const builtin = @import("builtin");
+const is_test = builtin.is_test;
+const build_options = @import("build_options");
+const mock_path = build_options.arch_mock_path;
+const arch = if (is_test) @import(mock_path ++ "arch_mock.zig") else @import("arch.zig");
+const log = if (is_test) @import(mock_path ++ "log_mock.zig") else @import("../../log.zig");
+const panic = if (is_test) @import(mock_path ++ "panic_mock.zig").panic else @import("../../panic.zig").panic;
+const gdt = @import("gdt.zig");
+const ComptimeBitmap = @import("../../bitmap.zig").ComptimeBitmap;
+const Allocator = std.mem.Allocator;
+
+/// The kernels main stack start as this is used to check for if the task being destroyed is this stack
+/// as we cannot deallocate this.
+extern var KERNEL_STACK_START: *u32;
+
+/// The function type for the entry point.
+const EntryPointFn = fn () void;
+
+/// The bitmap type for the PIDs
+const PidBitmap = if (is_test) ComptimeBitmap(u128) else ComptimeBitmap(u1024);
+
+/// The default stack size of a task. Currently this is set to a page size.
+const STACK_SIZE: u32 = arch.MEMORY_BLOCK_SIZE / @sizeOf(u32);
+
+/// The list of PIDs that have been allocated.
+var all_pids: PidBitmap = brk: {
+    var pids = PidBitmap.init();
+    // Set the first PID as this is for the current task running, init 0
+    _ = pids.setFirstFree() orelse unreachable;
+    break :brk pids;
+};
+
+/// The task control block for storing all the information needed to save and restore a task.
+pub const Task = struct {
+    const Self = @This();
+
+    /// The unique task identifier
+    pid: PidBitmap.IndexType,
+
+    /// Pointer to the stack for the task. This will be allocated on initialisation.
+    stack: []u32,
+
+    /// The current stack pointer into the stack.
+    stack_pointer: usize,
+
+    ///
+    /// Initialise a task. This will allocate a PID and the stack. The stack will be set up as a
+    /// kernel task. As this is a new task, the stack will need to be initialised with the CPU
+    /// state as described in arch.CpuState struct.
+    ///
+    /// Arguments:
+    ///     IN entry_point: EntryPointFn - The entry point into the task. This must be a function.
+    ///     IN allocator: *Allocator     - The allocator for allocating memory for a task.
+    ///
+    /// Return: *Task
+    ///     Pointer to an allocated task. This will then need to be added to the task queue.
+    ///
+    /// Error: Allocator.Error
+    ///     OutOfMemory - If there is no more memory to allocate. Any memory or PID allocated will
+    ///                   be freed on return.
+    ///
+    pub fn init(entry_point: EntryPointFn, allocator: *Allocator) Allocator.Error!*Task {
```

Comment:
Could this return a `Task` rather than a `*Task`? Then it's up to the caller to decide if it should go on the stack or heap. Makes it a bit more flexible.

Reply:
But when would you want to put a task on the stack as most likely the function called that creates a task will go out of scope so will be safer if it is in allocated memory

Reply:
True. I guess we could change it if the necessity arises.